hledger-lib/Hledger/Read/Common.hs

```diff
@@ -273,11 +286,24 @@ parseAndFinaliseJournal' parser iopts f txt = do
   case ep of
     Left e   -> throwError $ customErrorBundlePretty e
 
-    Right pj -> 
-      let pj' = if auto_ iopts then applyTransactionModifiers pj else pj in
-      case journalFinalise t f txt (not $ ignore_assertions_ iopts) pj' of
-                        Right j -> return j
-                        Left e  -> throwError e
+    Right pj ->
+      -- If we are using automated transactions, we finalize twice:
+      -- once before and once after. However, if we are running it
+      -- twice, we don't check assertions the first time (they might
+      -- be false pending modifiers) and we don't reorder the second
+      -- time. If we are only running once, we reorder and follow the
+      -- options for checking assertions.
+      let runFin :: Bool -> Bool -> (ParsedJournal -> Either String Journal)
+          runFin reorder ignore = journalFinalise t f txt reorder ignore
+          fj = if auto_ iopts
+               then applyTransactionModifiers <$>
+                    runFin True False pj >>=
+                    runFin False (not $ ignore_assertions_ iopts)
+               else runFin True (not $ ignore_assertions_ iopts) pj
+      in
+        case fj of
+          Right j -> return j
+          Left e  -> throwError e
```

Comment:
IIRC this code is already a bit hard to grok, so I don't love that we are adding more complexity. 

The commit doc is clearer than the code comment, perhaps you can move more of it into the code (as haddock or ordinary comment).

I really don't love the idea of duplicating this twice. Or that we have these two similar looking functions, one with no docs. I wonder if we can clean this up.